### PR TITLE
Fixing linecache

### PR DESCRIPTION
**Pull request by sjoshid**
@linde12 
Here's my attempt to fixing this. It might not be optimal as Im a Go beginner. I have referred to xrl crate which is used by xterm. 
There are still couple of issues I see. 2 important ones are:
1) 'update' case in linecache.go has not changed. Not sure how I can test this piece of code. No matter what I do in editor, 'update' op is never returned by core. Leaving this as is. 
2) If you move cursor to end of last line and hit return, the cursor doesnt move to next line. But if you start typing it enters the line in correct location. 

Other issues (like line numbers in my viewport doesnt update as I expand my current window) I think are unrelated to linecache. But I may be wrong.

### CODE REVIEW

**Comment by linde12** [comment]
Hey @sjoshid glad to see a patch for this. I've tried it out and it doesn't seem to work for me. I still get the same behavior as described in #7 (your gif) using the latest xi-core and your branch.

The update case should be triggered by simply moving the cursor. The following is a log (out.log) of me moving the cursor up two lines:

```
2018/04/13 13:15:10 >>> {"method":"edit","params":{"method":"move_up","params":{},"view_id":"view-id-1"},"view_id":"view-id-1"}
2018/04/13 13:15:10 <<< {"method":"update","params":{"update":{"ops":[{"n":42,"op":"copy"},{"lines":[{"cursor":[0],"styles":[],"text":"}\n"},{"styles":[],"text":""}],"n":2,"op":"ins"}],"pristine":true},"view_id":"view-id-1"}}
2018/04/13 13:15:10 <<< {"method":"scroll_to","params":{"col":0,"line":42,"view_id":"view-id-1"}}
2018/04/13 13:15:10 unhandled request: scroll_to
2018/04/13 13:15:10 >>> {"method":"edit","params":{"method":"move_up","params":{},"view_id":"view-id-1"},"view_id":"view-id-1"}
2018/04/13 13:15:10 <<< {"method":"update","params":{"update":{"ops":[{"n":41,"op":"copy"},{"lines":[{"cursor":[0],"styles":[],"text":"\te.Start()\n"},{"styles":[],"text":"}\n"}],"n":2,"op":"ins"},{"n":2,"op":"skip"},{"n":1,"op":"copy"}],"pristine":true},"view_id":"view-id-1"}}
2018/04/13 13:15:10 <<< {"method":"scroll_to","params":{"col":0,"line":41,"view_id":"view-id-1"}}
2018/04/13 13:15:10 unhandled request: scroll_to
```

So i'm guessing it's the update that is causing the problem. Are you running an older version of the core, maybe that's why it's working for you?

**Comment by sjoshid** [comment]
Hi @linde12 
It seems to be working for me. Here's another [gif](https://giphy.com/gifs/5b9hDh6ylKoQFUJgg9)
I have the latest core. 

And what you are referring to when you move cursor is not the `update` operation. The `ops` array in json should have an `update` op just like `copy`, `insert`, etc. There is `select case` in linecache.go which has an `update` as one of its cases. Im talking about this `update` case. 
Please correct me if Im wrong.

**Comment by linde12** [comment]
Okay, maybe i'm doing something wrong. Let me try again tomorrow or on monday!

**Comment by linde12** [comment]
@sjoshid 

OK. I've tried again and it still won't work for me. I grabbed the latest xi-editor master (google/xi-editor@908819c205b1e190a230203cb24c47e3eef0ed10) combined with your master (1bb2e60822fcee661fd53bda5d8bca8b3b590c8b)

**Comment by sjoshid** [comment]
@linde12 
Ok. Sorry it took me this long to reply. Go is making me go nuts!! 
But I think I know what might be going on. Let me start by saying you are correct. Even though you have my master, it still wont work. Problem is, even though it's my branch, `kod.go` still imports from your source. The line causing this issue is `github.com/linde12/kod/editor`. When you do `go build` it is going to pick up your version of `linecache.go` which it shouldn't. Hence you get the same error. 
I have made one more commit https://github.com/linde12/kod/pull/11/commits/d75a1ac4de485e9ad208a579ba53c2d0b0fec6fc 
Of course, this shouldnt go in your master branch but at this point I'm not sure how to provide you with a working linecache. 
Give it a try. I hope it works.

**Comment by linde12** [comment]
@sjoshid 
Ah of course, silly of me. Go uses only absolute paths so your last commit doesn't build for me but i added your fork as a remote to the original repo and built from there and that worked. Great improvement :+1: 

```
~/s/g/s/g/l/kod (master) » git remote add sjoshid https://github.com/sjoshid/kod
~/s/g/s/g/l/kod (master) » git fetch --all
~/s/g/s/g/l/kod (master) » git checkout HEAD~
~/s/g/s/g/l/kod ((1bb2e60...)) » go build
```

Number two of your list of issues can be made into it's own issue, as this has value in itself :clap: Great work & thanks for the help.

File: editor/linecache.go [hunk 1]
```diff
@@ -50,28 +48,62 @@ func (lc *LineCache) ApplyUpdate(update *rpc.Update) {
...
-				op.N--
-				index++
+
+			if op.N < len(lc.lines) {
+				firstNLines := append(lc.lines[:0], lc.lines[:op.N]...)
+				lc.lines = lc.lines[op.N:]
+				newLines = append(newLines, firstNLines...) //remove from lc.lines
+				continue
+			} else {
+				//allNLines := append(lc.lines[:0], lc.lines[0:]...)
```
Comment (linde12) [inline editor/linecache.go]: think we can just remove this comment

File: editor/linecache.go [hunk 2]
```diff
@@ -50,28 +48,62 @@ func (lc *LineCache) ApplyUpdate(update *rpc.Update) {
...
 		case "invalidate":
 			lc.addInvalid(newLines, &newInvalidBefore, &newInvalidAfter, op.N)
 		case "ins":
 			for _, line := range op.Lines {
 				newline := NewLine(line.Text, line.Cursor, line.Styles)
-				newLines = lc.addLine(newLines, &newInvalidBefore, &newInvalidAfter, newline)
+				newLines = append(newLines, newline)
 			}
 		case "update":
 			for _, line := range op.Lines {
```
Comment (linde12) [inline editor/linecache.go:108]: Not entirely sure when the update case is used, but i can imagine it would be called by third party plugins. This can be a separate issue as well.

**Review by linde12 — CHANGES_REQUESTED** [review]
just remove that comment and we should be good, nice job

**Comment by sjoshid** [comment]
Welcome @linde12 
Thanks for bearing with Go newbie. I have removed those comments.

File: editor/linecache.go (hunk 2 above)
Comment (cmyr) [reply editor/linecache.go:108]: Update is intended as an optimization: if only selections or cursor position has changed, they can be updated without invalidating the whole line.

I haven't looked at this code in a while, but I'm pretty sure it's currently unused.

**Comment by linde12** [comment]
No sign of you being a newbie here @sjoshid  :smiley: Thanks a lot!
